### PR TITLE
#167165062 Add User Rating Feature

### PR DESCRIPTION
**Pull request by mekzy-o**
### What does this PR do?

- Adds functionality for users to rate articles
- Adds validation for rate value

#### Description of Task to be completed?
Users should be able to Rate an article
Rate articles - `POST /api/v1/articles/ratings`

#### How should this be manually tested?
After cloning this repo, `cd` into it and do the following:

```bash
# fetch this branch
$ git fetch origin  ft-user-create-article-and-tags-167165061-167165069

# install dependencies
$ npm install

# run unit tests
$ npm test

# start the development server
$ npm run server
```
Using postman, log in or register an account and hit the post route to rate an article (user must be verified via the verification route in #28 

add the following format in the request body:

```JSON
{
 "id": "ffff-e8790-76543563",
  "rate": 5,
}
```

#### Any background context you want to provide?
N/a

#### What are the relevant pivotal tracker stories?
[#167165062](https://www.pivotaltracker.com/story/show/167165062) 

#### Screenshots (if appropriate)

*Rate article successfully*
Screenshot 2019-08-01 at 8 59 26 AM

### CODE REVIEW

**Comment by VictorMaria** [comment]
Hello Emeka, you should consider rebasing your branch. Well done.

File: __tests__/mockData/mockRates.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// import uuid from 'uuid/v4';
```
Comment (tolulope-od) [inline __tests__/mockData/mockRates.js]: Hey Emeka, why is the code here commented out? Wouldn't it be better to just delete it?

File: __tests__/rateRoutes.test.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        });
+    });
+    it('should throw error when author tries to rate his article', (done) => {
+      chai
+        .request(app)
+        .post(`${url}`)
+        .set('Authorization', `Bearer ${validToken}`)
+        .send(mockRates[1])
+        .end((err, res) => {
+          expect(res.status).to.equal(400);
```
Comment (tolulope-od) [inline __tests__/rateRoutes.test.js]: Great implementation! I like that you've considered the edge case where an author tries to rate his own article

**Review by tolulope-od — COMMENTED** [review]
Good job here Emeka, I like the way you abstracted the helpers to make your code a lot neater and logical.

Please refer to my other comments on the PR and let me know if you have some thoughts. Thanks

File: __tests__/mockData/mockRates.js (hunk 1 above)
Comment (mekzy-o) [reply __tests__/mockData/mockRates.js]: Thanks. I have already changed that

File: __tests__/ratingRoutes.test.js [hunk 3]
```diff
@@ -57,4 +83,101 @@ describe('Check ratings', () => {
...
+        .end((err, res) => {
+          expect(res.status).to.equal(201);
+          const { article } = res.body;
+          expect(article.message).to.be.eql('Your rating has been added successfully');
+          expect(article).to.have.property('articleId');
+          expect(article).to.have.property('rate');
+          expect(article).to.have.property('firstName');
+          expect(article).to.have.property('lastName');
+          done();
+        });
```
Comment (mazma1) [inline __tests__/ratingRoutes.test.js:118]: I've given this feedback about writing specific assertions for our tests. Please fix this for all tests where you missed it

File: __tests__/ratingRoutes.test.js [hunk 4]
```diff
@@ -57,4 +83,101 @@ describe('Check ratings', () => {
...
+        .set('Authorization', `Bearer ${verifiedUserToken}`)
+        .send(mockRates[1])
+        .end((err, res) => {
+          expect(res.status).to.equal(400);
+          expect(res.body.errors.message).to.be.eql("You can't rate your article");
+          done();
+        });
+    });
+    it('should update rating when author tries to rate an article again', (done) => {
+      chai
```
Comment (mazma1) [inline __tests__/ratingRoutes.test.js:133]: Should this be for an author rating an article, or a user?

Comment (mekzy-o) [reply __tests__/ratingRoutes.test.js:133]: Okay, I'll change that

File: __tests__/ratingRoutes.test.js (hunk 3 above)
Comment (mekzy-o) [reply __tests__/ratingRoutes.test.js:118]: Okay I would. Thanks

Modified: I just checked the test, I'm sending only the user login details(email & password).. The first Name and lastName are the property of the response. the request object doesn't have it

Comment (mazma1) [reply __tests__/ratingRoutes.test.js:118]: I'm sure `seededUsers[0]` has a first and last name which is the specific value returned in the response.

How about the rating you submitted, does that not have a specific value?

File: __tests__/ratingRoutes.test.js [hunk 5]
```diff
@@ -57,4 +83,101 @@ describe('Check ratings', () => {
...
+        .set('Authorization', `Bearer ${verifiedUserToken}`)
+        .send(mockRates[2])
+        .end((err, res) => {
+          const { article } = res.body;
+          expect(res.status).to.equal(201);
+          expect(article.message).to.be.eql('Your rating has been added successfully');
+          expect(article).to.have.property('articleId');
+          expect(article).to.have.property('rate');
+          done();
+        });
```
Comment (mazma1) [inline __tests__/ratingRoutes.test.js:146]: Fix this test case too

File: __tests__/ratingRoutes.test.js (hunk 3 above)
Comment (mekzy-o) [reply __tests__/ratingRoutes.test.js:118]: seederUsers[0] is just the login details of the user. It could be any user. 

I have fixed the assertion for the rating value I submitted
